### PR TITLE
Less strict code of the TinylogLoggingProvider #168

### PR DESCRIPTION
**Pull request by Git5000**
### Description

Implementation of the changes discussed in issue #168.

- Moved the writer generation code from `ConfigurationParser` into the non static, non final class `TinylogLoggingConfiguration` to allow other providers to use it.
- Moved commonly useful methods like log entry handling from `TinylogLoggingProvider` into the non static, non final class `TinylogLoggingConfiguration`
- Contrary to the initial suggestion did not change the static code of `ConfigurationParser` as it now just calls static methods of Configuration. So static is ok.
- Some smaller methods are kept in `TinylogLoggingProvider` (like level calculation) as this will be used inside the logging write  itself and might affect the performance (all other changes are constructor only changes). Can be discussed but I think it is ok to keep it in its original place.
- Added new unit tests to cover these changes
- It could be discussed as idea (but I decided not to follow it for the moment) to put other access function to `Configuration` elements (e.g. filename) which is done by the writers constructors or support methods into `ConfigurationParser`  as the parser now more or less translates string elements of `Configuration`  into method getters. Similar functions might be `getFilename()`  and alike from the writers. I did not follow this up as it can also be argued that there is not so much common or usefull stuff to justify it and the writers know themself best what they are doing. So here just as possibility.

As result you can now more easily duplicate/write an own logging provider with similar functionality to `TinylogLoggingProvider` and use the helper functions from `TinylogLoggingConfiguration`. Much less code duplication is now needed.

Linked issue: #168 

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including corner cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):

```markdown
No changed to documentation.
```

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt)
- [x] I agree that my GitHub user name will be published in the release notes

### CODE REVIEW

**Comment by Git5000** [comment]
Please check the pull request throurougly. It is a bit more code change so you should look at it whether it is all ok or tell me where you desire changes.

The change can go to 2.2. or 2.3. 

It does not change API or usage and the tests pass but of course it is a some LOC changes  which always can introduce undetected bugs - although it is mainly moving the methods to another file.

**Comment by Git5000** [comment]
PS: The latest top level POM file has the javadoc line 
`<javadocExecutable>${java.home}/bin/javadoc</javadocExecutable>`
 removed.  Without this it does not build with me anymore. What do I have to change to build with the latest POM?

File: tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+import org.tinylog.Level;
+import org.tinylog.Supplier;
+import org.tinylog.configuration.Configuration;
+import org.tinylog.configuration.ServiceLoader;
+import org.tinylog.format.MessageFormatter;
+import org.tinylog.runtime.RuntimeProvider;
+import org.tinylog.runtime.Timestamp;
+import org.tinylog.writers.Writer;
+
+public class TinylogLoggingConfiguration {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingConfiguration.java:36]: Please add a Javadoc comment

File: tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	public TinylogLoggingConfiguration() {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingConfiguration.java:37]: Please put a blank line between class declaration and constructor

File: tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingProvider.java [hunk 3]
```diff
@@ -145,7 +141,8 @@ public void log(final int depth, final String tag, final Level level, final Thro
 		}
 
 		if (activeLevel.ordinal() <= level.ordinal()) {
-			LogEntry logEntry = createLogEntry(stackTraceElement, tag, tagIndex, level, exception, formatter, obj, arguments);
+			LogEntry logEntry = TinylogLoggingConfiguration.createLogEntry(stackTraceElement, tag, level, exception, formatter, 
+																			obj, arguments, requiredLogEntryValues[tagIndex], context);
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingProvider.java]: Something odd with indentation?

File: tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingProvider.java [hunk 4]
```diff
@@ -176,15 +173,16 @@ public void log(final String loggerClassName, final String tag, final Level leve
 		}
 
 		if (activeLevel.ordinal() <= level.ordinal()) {
-			LogEntry logEntry = createLogEntry(stackTraceElement, tag, tagIndex, level, exception, formatter, obj, arguments);
+			LogEntry logEntry = TinylogLoggingConfiguration.createLogEntry(stackTraceElement, tag, level, exception, formatter,  
+																			obj, arguments, requiredLogEntryValues[tagIndex], context);
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/core/TinylogLoggingProvider.java]: Something odd with indentation?

File: tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+		assertThat(requirements.get(4)).isTrue();
+	}	
+
+	/**
+	 * Verifies that all writers can be obained.
+	 */
+	@Test
+	@SuppressWarnings("unchecked")
+	public void getAllWriters() {
+
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java:382]: Please remove blank line

File: tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+			.hasAtLeastOneElementOfType(WriterTwo.class);
+	}		
+	
+	/**
+	 * Verifies that a log entry can be created properly.
+	 */
+	@Test
+	@SuppressWarnings("unchecked")
+	public void createLogEntry() {
+
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java:404]: Please remove blank line

File: tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+		requiredLogEntryValues[Level.INFO.ordinal()] = Arrays.asList(new LogEntryValue[] {LogEntryValue.CONTEXT});
+		LogEntry entry2 = TinylogLoggingConfiguration.createLogEntry(null, null, Level.INFO, null, null, null,
+																		null, requiredLogEntryValues, contextProvider);
+		assertThat(entry2.getContext()).containsAllEntriesOf(Collections.singletonMap("id", "100"));
+		
+		requiredLogEntryValues[Level.INFO.ordinal()] = Arrays.asList(new LogEntryValue[] {LogEntryValue.EXCEPTION});
+		LogEntry entry3 = TinylogLoggingConfiguration.createLogEntry(null, null, Level.INFO, new Exception("test"), null, null,
+																		null, requiredLogEntryValues, null);
+		assertThat(entry3.getException().getMessage()).isEqualTo("test");
+
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java:424]: Please remove blank line

File: tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+	@SuppressWarnings("unchecked")
+	public void createLogEntry() {
+		
+		TinylogContextProvider contextProvider = new TinylogContextProvider();
+		contextProvider.put("id", "100");
+		Collection<LogEntryValue>[] requiredLogEntryValues = new Collection[Level.values().length - 1];
+
+		requiredLogEntryValues[Level.INFO.ordinal()] = Arrays.asList(new LogEntryValue[] {LogEntryValue.TAG, LogEntryValue.LEVEL});
+		LogEntry entry1 = TinylogLoggingConfiguration.createLogEntry(null, "testtag", Level.INFO, null, null, null,
+																		null, requiredLogEntryValues, null);
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/core/TinylogLoggingConfigurationTest.java:411]: Something odd with indentation (also twice below)?

**Review by pmwmedia — CHANGES_REQUESTED** [review]
Thank you very much for your great pull request! It really makes it easier to implement custom logging providers. I have found only a few minor style issues during my review.

> The latest top level POM file has the javadoc line <javadocExecutable>${java.home}/bin/javadoc</javadocExecutable> removed.

Actually, Javadoc generation makes a lot of trouble. On some JDKs, the Javadoc generation works only with this line and on others, only without. I couldn't find any good solution for this problem yet.

I have noticed a merge in the commits. I think we should try to avoid this merge. Could you try to checkout the latest branch v2.3 (I have just created it), copy manually your five changed file there, and commit them to the new branch?

**Comment by Git5000** [comment]
You are right a bit odd indentation...  Must be some editor settings.  

I will fix all style errors and create the pull request again against 2.3. as you suggested tomorrow.

**Comment by pmwmedia** [comment]
Closed in favor of #173
